File: 416.partition.equal.subset.sum.c
bool canPartition(int* nums, int numsSize) {
    if (numsSize <=0) return false;
    int max = numsSize * 100;
    bool *dp = malloc(max * sizeof(bool));
    memset(dp, 0, max * sizeof(bool));
    dp[0] = true;
    bool ret = false;
    int sum = 0;
    for(int i=0;i<numsSize; ++i)
    {
        sum += nums[i];
        for(int j=max-1; j>=nums[i]; --j)
        {
            dp[j] |= dp[j-nums[i]];
        }
    }
    if (!(sum&1))
    {
        ret = dp[sum>>1];
    }
    free(dp);
    return ret;
}

File: 416. Partition Equal Subset Sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        n = len(nums)
        total_sum = sum(nums)
        if n < 2 or total_sum & 1 == 1:
            return False
        
        target = total_sum // 2
        subsets = set([0])
        for num in nums:
            new_subsets = []
            for sub in subsets:
                new_sub = sub + num
                if new_sub == target:
                    return True
                if new_sub not in subsets and new_sub < target:
                    new_subsets.append(new_sub)
            subsets.update(new_subsets)
        return False
            
